### PR TITLE
`eval_spec_value` and `eval_prim`

### PR DESCRIPTION
**Pull request by cdisselkoen**
_Draft for now due to including `sorry`s._

This is another lemma for the PE soundness proof.  It says that partial evaluation of a `Spec.Value` always succeeds and returns the value.  This PR also includes some simple corollaries.

### CODE REVIEW

File: cedar-lean/Cedar/Thm/Ext.lean [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  simp [Spec.call, Spec.res, Coe.coe]
+  cases h₁ : IPAddr.ip (toString ip)
+  <;> simp only [Except.ok.injEq, Spec.Value.ext.injEq, ipaddr.injEq]
+  case none =>
+    simp [IPAddr.ip, IPAddr.parse] at h₁
+    split at h₁ <;> rename_i h₂
+    · replace ⟨ipnet, h₂⟩ := Option.isSome_iff_exists.mp h₂
+      simp [h₂] at h₁
+    · simp only [Option.not_isSome, Option.isNone_iff_eq_none, Bool.not_eq_true] at h₂
+      sorry -- can't unfold these private definitions; can we make them not private?
```
Comment (emina) [inline cedar-lean/Cedar/Thm/Ext.lean:51]: Yes, we'll need to make these definitions non-private to unfold them.

File: cedar-lean/Cedar/Thm/Partial/Evaluation/Record.lean [hunk 2]
```diff
@@ -182,6 +182,93 @@ theorem partial_eval_wf {attrs: List (Attr × Partial.Expr)} {request : Partial.
...
       simpa [Partial.Value.WellFormed] using ih (k, v) h₅ (.value v') h₇
 
+private theorem mapM_Except_on_snd_preserves_sortedBy_fst [LT α] {abs: List (α × β)} {f : β → Except ε γ} :
+  abs.SortedBy Prod.fst →
+  abs.mapM (λ (a, b) => do let b' ← f b ; .ok (a, b')) = .ok ags →
+  ags.SortedBy Prod.fst
+:= by
+  sorry
+
+private theorem mapM_Option_on_snd_preserves_sortedBy_fst [LT α] {abs : List (α × β)} {f : β → Option γ} :
```
Comment (emina) [inline cedar-lean/Cedar/Thm/Partial/Evaluation/Record.lean:192]:
```suggestion
private theorem mapM_Option_on_snd_preserves_sortedBy_fst [LT α] [DecidableLT α] [StrictLT α] {abs : List (α × β)} {f : β → Option γ} :
  abs.SortedBy Prod.fst →
  abs.mapM (λ (a, b) => do some (a, ← f b)) = some ags →
  ags.SortedBy Prod.fst
:= by
  intro h₁ h₂ 
  replace h₂ := List.mapM_some_eq_filterMap h₂
  subst h₂
  apply List.filterMap_sortedBy _ h₁ 
  simp only [Option.bind_eq_bind, Option.bind_eq_some, Option.some.injEq, forall_exists_index,
    and_imp, forall_apply_eq_imp_iff₂, implies_true]
```

File: cedar-lean/Cedar/Thm/Partial/Evaluation/Record.lean [hunk 3]
```diff
@@ -182,6 +182,93 @@ theorem partial_eval_wf {attrs: List (Attr × Partial.Expr)} {request : Partial.
       subst k' pval'
       simpa [Partial.Value.WellFormed] using ih (k, v) h₅ (.value v') h₇
 
+private theorem mapM_Except_on_snd_preserves_sortedBy_fst [LT α] {abs: List (α × β)} {f : β → Except ε γ} :
```
Comment (emina) [inline cedar-lean/Cedar/Thm/Partial/Evaluation/Record.lean:185]: Proof similar to `mapM_Option_on_snd_preserves_sortedBy_fst`. 

We can write the body of the lambda more compactly as `(λ (a, b) => do .ok (a, ← f b))`.  Lean will take care of introducing the intermediate variables.

Comment (cdisselkoen) [reply cedar-lean/Cedar/Thm/Partial/Evaluation/Record.lean:185]: that's awesome syntax sugar

File: cedar-lean/Cedar/Thm/Partial/Evaluation/Record.lean (hunk 2 above)
Comment (cdisselkoen) [reply cedar-lean/Cedar/Thm/Partial/Evaluation/Record.lean:192]: Thanks!

**Comment by cdisselkoen** [comment]
This lemma is no longer needed after #379.  Closing this.

**Pull request closed without merging**